tools/adventure-pack/src/__tests__/iteratorPrototypeFilter.test.ts

```diff
@@ -0,0 +1,64 @@
+import { describe, expect, it } from "@jest/globals";
+
+import "../iteratorPrototypeFilter";
+
+describe("Iterator.prototype.filter", () => {
+  it("can filter an Array's values()", () => {
+    const array = [2, 4, 6, 7, 8];
+
+    const filterResult = array.values().filter((p) => p % 2 === 0);
+    expect(filterResult.next().value).toBe(2);
+    expect(filterResult.next().value).toBe(4);
+    expect(filterResult.next().value).toBe(6);
+    expect(filterResult.next().value).toBe(8);
+    expect(filterResult.next().done).toBe(true);
+  });
+
+  it("can filter an Array's values() with different condition", () => {
+    const array = [1, 2, 3, 4, 5];
```

Comment:
When writing tests, I think it's good to think a bit like how can I make sure the test doesn't accidentally pass?

For example, let's imagine an implementation that looks something like this:

```ts
let index = 0;
let foundElementThatPassesFilter = false;
for (const element of iteratorToIterable(this)) {
  if (callbackFn(element, index)) {
   foundElementThatPassesFilter = true;
  }
  if (foundElementThatPassesFilter) {
    yield element;
  }
  ++index;
}
```

In other words, it filters until the first match, then yields everything that follows.

This test case won't identify that this implementation is flawed.

Similarly, if some implementation filtered on the array index instead of the array value for whatever reason, this test might also not catch it, since the array values are so close to being exactly the same as the indexes.

Obviously it's easy to get super paranoid, and I don't think we need to go overboard with it, but I'd recommend slightly stranger test cases that are less likely to accidentally pass.

Reply:
thanks this is really helpful while I'm writing out test!